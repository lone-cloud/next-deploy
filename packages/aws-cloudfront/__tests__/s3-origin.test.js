const {
  mockCreateDistribution,
  mockUpdateDistribution,
  mockCreateDistributionPromise,
  mockGetDistributionConfigPromise,
  mockUpdateDistributionPromise,
  mockCreateCloudFrontOriginAccessIdentityPromise,
  mockPutBucketPolicy,
} = require('aws-sdk');

const { createComponent, assertHasOrigin } = require('../test-utils');

jest.mock('aws-sdk', () => require('../__mocks__/aws-sdk.mock'));

describe('S3 origins', () => {
  let component;

  beforeEach(async () => {
    mockCreateDistributionPromise.mockResolvedValueOnce({
      Distribution: {
        Id: 'distributionwithS3origin',
      },
    });

    component = await createComponent();
  });

  describe('When origin is an S3 bucket URL', () => {
    it('creates distribution', async () => {
      await component.default({
        origins: ['https://mybucket.s3.amazonaws.com'],
      });

      assertHasOrigin(mockCreateDistribution, {
        Id: 'mybucket',
        DomainName: 'mybucket.s3.amazonaws.com',
        S3OriginConfig: {
          OriginAccessIdentity: '',
        },
        CustomHeaders: {
          Quantity: 0,
          Items: [],
        },
        OriginPath: '',
      });

      expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot();
    });

    it('updates distribution', async () => {
      mockGetDistributionConfigPromise.mockResolvedValueOnce({
        ETag: 'etag',
        DistributionConfig: {
          Origins: {
            Items: [],
          },
        },
      });
      mockUpdateDistributionPromise.mockResolvedValueOnce({
        Distribution: {
          Id: 'distributionwithS3originupdated',
        },
      });

      await component.default({
        origins: ['https://mybucket.s3.amazonaws.com'],
      });

      await component.default({
        origins: ['https://anotherbucket.s3.amazonaws.com'],
      });

      assertHasOrigin(mockUpdateDistribution, {
        Id: 'anotherbucket',
        DomainName: 'anotherbucket.s3.amazonaws.com',
      });

      expect(mockUpdateDistribution.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('When origin is an S3 URL only accessible via CloudFront', () => {
    it('creates distribution', async () => {
      mockCreateCloudFrontOriginAccessIdentityPromise.mockResolvedValueOnce({
        CloudFrontOriginAccessIdentity: {
          Id: 'access-identity-xyz',
          S3CanonicalUserId: 's3-canonical-user-id-xyz',
        },
      });

      await component.default({
        origins: [
          {
            url: 'https://mybucket.s3.amazonaws.com',
            private: true,
          },
        ],
      });

      expect(mockPutBucketPolicy).toBeCalledWith({
        Bucket: 'mybucket',
        Policy: expect.stringContaining('"CanonicalUser":"s3-canonical-user-id-xyz"'),
      });

      assertHasOrigin(mockCreateDistribution, {
        Id: 'mybucket',
        DomainName: 'mybucket.s3.amazonaws.com',
        S3OriginConfig: {
          OriginAccessIdentity: 'origin-access-identity/cloudfront/access-identity-xyz',
        },
        CustomHeaders: {
          Quantity: 0,
          Items: [],
        },
        OriginPath: '',
      });

      expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot();
    });

    it('updates distribution', async () => {
      mockCreateCloudFrontOriginAccessIdentityPromise.mockResolvedValue({
        CloudFrontOriginAccessIdentity: {
          Id: 'access-identity-xyz',
          S3CanonicalUserId: 's3-canonical-user-id-xyz',
        },
      });

      mockGetDistributionConfigPromise.mockResolvedValueOnce({
        ETag: 'etag',
        DistributionConfig: {
          Origins: {
            Items: [],
          },
        },
      });

      mockUpdateDistributionPromise.mockResolvedValueOnce({
        Distribution: {
          Id: 'distributionwithS3originupdated',
        },
      });

      await component.default({
        origins: [
          {
            url: 'https://mybucket.s3.amazonaws.com',
            private: true,
          },
        ],
      });

      await component.default({
        origins: [
          {
            url: 'https://anotherbucket.s3.amazonaws.com',
            private: true,
          },
        ],
      });

      expect(mockPutBucketPolicy).toBeCalledWith({
        Bucket: 'anotherbucket',
        Policy: expect.stringContaining('"CanonicalUser":"s3-canonical-user-id-xyz"'),
      });

      assertHasOrigin(mockUpdateDistribution, {
        Id: 'anotherbucket',
        DomainName: 'anotherbucket.s3.amazonaws.com',
        S3OriginConfig: {
          OriginAccessIdentity: 'origin-access-identity/cloudfront/access-identity-xyz',
        },
        OriginPath: '',
        CustomHeaders: { Items: [], Quantity: 0 },
      });

      expect(mockCreateDistribution.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
