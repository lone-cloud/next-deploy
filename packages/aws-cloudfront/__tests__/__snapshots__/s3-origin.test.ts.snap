// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3 origins When origin is an S3 URL only accessible via CloudFront creates distribution 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CacheBehaviors": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "mybucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "mybucket.s3.amazonaws.com",
          "Id": "mybucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "origin-access-identity/cloudfront/access-identity-xyz",
          },
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`S3 origins When origin is an S3 URL only accessible via CloudFront updates distribution 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CacheBehaviors": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "mybucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "mybucket.s3.amazonaws.com",
          "Id": "mybucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "origin-access-identity/cloudfront/access-identity-xyz",
          },
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`S3 origins When origin is an S3 bucket URL creates distribution 1`] = `
Object {
  "DistributionConfig": Object {
    "Aliases": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CacheBehaviors": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "CallerReference": "1566599541192",
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "mybucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "HttpVersion": "http2",
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "mybucket.s3.amazonaws.com",
          "Id": "mybucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "",
          },
        },
      ],
      "Quantity": 1,
    },
    "PriceClass": "PriceClass_All",
  },
}
`;

exports[`S3 origins When origin is an S3 bucket URL updates distribution 1`] = `
Object {
  "DistributionConfig": Object {
    "CacheBehaviors": Object {
      "Items": Array [],
      "Quantity": 0,
    },
    "Comment": "",
    "DefaultCacheBehavior": Object {
      "AllowedMethods": Object {
        "CachedMethods": Object {
          "Items": Array [
            "HEAD",
            "GET",
          ],
          "Quantity": 2,
        },
        "Items": Array [
          "HEAD",
          "GET",
        ],
        "Quantity": 2,
      },
      "Compress": false,
      "DefaultTTL": 86400,
      "FieldLevelEncryptionId": "",
      "ForwardedValues": Object {
        "Cookies": Object {
          "Forward": "none",
        },
        "Headers": Object {
          "Items": Array [],
          "Quantity": 0,
        },
        "QueryString": false,
        "QueryStringCacheKeys": Object {
          "Items": Array [],
          "Quantity": 0,
        },
      },
      "LambdaFunctionAssociations": Object {
        "Items": Array [],
        "Quantity": 0,
      },
      "MaxTTL": 31536000,
      "MinTTL": 0,
      "SmoothStreaming": false,
      "TargetOriginId": "anotherbucket",
      "TrustedSigners": Object {
        "Enabled": false,
        "Items": Array [],
        "Quantity": 0,
      },
      "ViewerProtocolPolicy": "redirect-to-https",
    },
    "Enabled": true,
    "Origins": Object {
      "Items": Array [
        Object {
          "CustomHeaders": Object {
            "Items": Array [],
            "Quantity": 0,
          },
          "DomainName": "anotherbucket.s3.amazonaws.com",
          "Id": "anotherbucket",
          "OriginPath": "",
          "S3OriginConfig": Object {
            "OriginAccessIdentity": "",
          },
        },
      ],
      "Quantity": 1,
    },
  },
  "Id": "distributionwithS3origin",
  "IfMatch": "etag",
}
`;
